Normative: Allow CodeLike object evaluation

lukewarlow opened the pull request:
This change introduces a new HostGetCodeForEval host hook, this is used to allow evaluating certain objects.

This change also provides the compilation type and the original arguments, to the host hook HostEnsureCanCompileStrings.

This PR upstreams ~~a reduced version of~~ the https://github.com/tc39/proposal-dynamic-code-brand-checks proposal 

Closes https://github.com/tc39/ecma262/pull/1498, https://github.com/tc39/ecma262/issues/938
<!--
If you are changing the signature or behavior of an existing construct, please check if this affects downstream dependencies (searching for the construct's name is sufficient) and if needed file an issue:

* [Web IDL](https://webidl.spec.whatwg.org/) — [file an issue](https://github.com/whatwg/webidl/issues/new)
* [HTML Standard](https://html.spec.whatwg.org/) — [file an issue](https://github.com/whatwg/html/issues/new)
* [ECMAScript Intl API](https://tc39.es/ecma402/) - [file an issue](https://github.com/tc39/ecma402/issues/new)

Note: please ensure that the "Allow edits and access to secrets by maintainers" checkbox remains checked.
-->

ljharb commented:
probably a little premature to do this at stage 1 :-)

caridy commented on spec.html:
```diff
@@ -30183,6 +30209,7 @@ <h1>
                 1. Set _k_ to _k_ + 1.
             1. Let _bodyParseString_ be the string-concatenation of 0x000A (LINE FEED), _bodyString_, and 0x000A (LINE FEED).
             1. Let _sourceString_ be the string-concatenation of _prefix_, *" anonymous("*, _P_, 0x000A (LINE FEED), *") {"*, _bodyParseString_, and *"}"*.
+            1. Perform ? HostEnsureCanCompileStrings(_currentRealm_, _parameterStrings_, _bodyString_, _sourceString_, ~function~, _parameterArgs_, _bodyArg_).
```
Can you elaborate more on why do we need to pass two strings for the body? _bodyString_ and _sourceString_ are different, but I'm not sure why would the Host need _bodyString_ that _sourceString_  and _bodyArg_ cannot provide?

lukewarlow replied:
Delegates decided against allowing the fully constructed codeString so this is no longer relevant.

caridy commented:
> probably a little premature to do this at stage 1 :-)

@ljharb I think the reasoning behind this PR is that it doesn't really change any functionality, it is just layering. The original proposal from @mikesamuel was a lot bigger, this is just a small piece of that and we were hoping to just ask for consensus rather than pushing this thru the staging process. @cc @nicolo-ribaudo

lukewarlow commented:
@ljharb my understanding was that this would just be layering in the same way https://github.com/tc39/ecma262/pull/3222 was.

ljharb commented:
You might be right :-) would this PR obviate the entire linked proposal?

lukewarlow commented:
Yeah with this PR the use case for that proposal trusted types protection of eval and Function would be entirely addressed. That proposal would subsequently be withdrawn. The changes have been reduced over time and that linked PR addressed some of the required changes anyway.

michaelficarra commented:
FYI @lukewarlow @nicolo-ribaudo this wouldn't be appropriate for stage 3 before testing and integration work has been completed. Maybe you should change the agenda item asking for stage 3 to stage 2.7?

michaelficarra commented on spec.html:
```diff
@@ -5738,6 +5738,23 @@ <h1>
+      </h1>
+      <dl class="header">
+        <dt>description</dt>
+        <dd>It checks if _argument_ contains the host-defined [[HostDefinedIsCodeLike]] internal slot.</dd>
```
It doesn't just check for the presence of the slot, it returns its value.

michaelficarra commented:
I don't really like the design here. Instead of adding a Boolean slot, could we add a String slot? The thing I don't like about having a Boolean slot and calling `ToString` on it is that there's no guarantee that it produces the same String each time. There's also no guarantee that `ToString` completes normally. If the String can't be precomputed and guaranteed constant, we could also use the slot to memoise it: `~not-computed~ or a String`.

If it doesn't limit what you're trying to do with this proposal, I would much prefer a design like that.

michaelficarra commented on spec.html:
```diff
@@ -28857,10 +28874,16 @@ <h1>
           1. Let _evalRealm_ be the current Realm Record.
           1. NOTE: In the case of a direct eval, _evalRealm_ is the realm of both the caller of `eval` and of the `eval` function itself.
-          1. Perform ? HostEnsureCanCompileStrings(_evalRealm_, « », _x_, _direct_).
+          1. Let _type_ be ~direct-eval~ if _direct_ is *true*, or ~indirect-eval~ if _direct_ is *false*.
```
I'd rather change the type of `_direct_` in `PerformEval`.

michaelficarra commented on spec.html:
```diff
@@ -28934,7 +28960,8 @@ <h1>
         <p>
           _parameterStrings_ represents the strings that, when using one of the function constructors, will be concatenated together to build the parameters list. _bodyString_ represents the function body or the string passed to an `eval` call.
-          _direct_ signifies whether the evaluation is a direct eval.
+          _codeString_ represents the full code string. _compilationType_ is an enum that indicates what type of compilation this is.
```
> \_compilationType\_ is an enum that indicates what type of compilation this is.

This sentence is nearly tautological. Either explain what each of the 3 possible values means or omit the sentence.

michaelficarra commented on spec.html:
```diff
@@ -28934,7 +28960,8 @@ <h1>
           _parameterStrings_ represents the strings that, when using one of the function constructors, will be concatenated together to build the parameters list. _bodyString_ represents the function body or the string passed to an `eval` call.
-          _direct_ signifies whether the evaluation is a direct eval.
+          _codeString_ represents the full code string. _compilationType_ is an enum that indicates what type of compilation this is.
+          _parameterArgs_ represents the list of values passed when using one of the function constructors. _bodyArg_ represents the function body or the value passed to an `eval` call.
```
```suggestion
          _parameterArgs_ are the values passed as the leading parameters to one of the Function constructors. _bodyArg_ is either the final parameter passed to on of the Function constructors or the value passed to an `eval` call.
```

nicolo-ribaudo commented:
@michaelficarra This is currently tested in WPT. I asked to @ptomato how to test this in test262 but, given that the only observable change is that there _might_ be objects that are not returned as-is by `eval()` but we don't actually say which objects they are (or we don't request that they exist), he suggested that this isn't really testable in test262 and that it's better to have it in WPT.

The test that shows that "there exist an object that is not returned as-is by `eval`" is at https://github.com/web-platform-tests/wpt/blob/23894a7e427bc088cfd4b3aa6ba8a8927b904713/trusted-types/eval-csp-no-tt.html#L13-L15. We could try somehow moving it to test262 if you prefer, maybe exposing a `$262.getObjectForEval(toStringResult)` function.

There is also a test about how it interacts with `new Function` (at https://github.com/web-platform-tests/wpt/blob/master/trusted-types/eval-function-constructor.html), but given that whether a call to `new Function` is allowed or not is entirely host-defined I don't think it can be moved to test262 in any way.

nicolo-ribaudo commented:
> I don't really like the design here. Instead of adding a Boolean slot, could we add a String slot? The thing I don't like about having a Boolean slot and calling `ToString` on it is that there's no guarantee that it produces the same String each time. There's also no guarantee that `ToString` completes normally. If the String can't be precomputed and guaranteed constant, we could also use the slot to memoise it: `~not-computed~ or a String`.
> 
> If it doesn't limit what you're trying to do with this proposal, I would much prefer a design like that.

I recommended using `toString()` to convert objects to string in `eval()` because `new Function` already uses `toString()` to convert objects to strings. When doing `new Function(someObject)` twice, there is already no guarantee that it will generate two functions with the same code.

When it comes to Trusted Types, the host is checking not only that the `new Function`/`eval` is being called with an object that has the capability of being converted into code, but also that the object gives the permission to convert _the specific string obtained through ecma262_ into code (step 2.4.1 of https://w3c.github.io/trusted-types/dist/spec/#host-ensure-can-compile-strings)

nicolo-ribaudo commented:
@syg I was reading https://github.com/tc39/proposal-source-phase-imports/issues/59 — should we replace the internal slot on the host-provided object with a `HostEvalShouldStringifyObject` hook for the same reason?

gibson042 commented on spec.html:
```diff
@@ -28857,10 +28874,16 @@ <h1>
         <emu-alg>
           1. Assert: If _direct_ is *false*, then _strictCaller_ is also *false*.
-          1. If _x_ is not a String, return _x_.
+          1. If _x_ is an Object and IsCodeLike(_x_) is *true*, then
+            1. Let _xStr_ be ? ToString(_x_).
```
Under what circumstances would this ToString throw? Should it instead be infallible?
```suggestion
          1. If _x_ is an Object and IsCodeLike(_x_) is *true*, then
            1. Let _xStr_ be ! ToString(_x_).
```

lukewarlow replied:
My understanding is that if an objects to string function throws this operation will too. It also appears to be `? ToString` in the Create dynamic function algorithm too.

gibson042 commented on spec.html:
```diff
@@ -28857,10 +28857,18 @@ <h1>
         <emu-alg>
           1. Assert: If _direct_ is *false*, then _strictCaller_ is also *false*.
-          1. If _x_ is not a String, return _x_.
+          1. If _x_ is an Object, then
+            1. Let _code_ be HostGetCodeForEval(_x_).
+            1. If _code_ is ~unknown~ return _x_.
+            1. Else, let _xStr_ be _code_.
+          1. Else if _x_ is a String, then
+            1. Let _xStr_ be _x_.
+          1. Else,
+            1. Return _x_.
```
Ordering nit:
```suggestion
          1. If _x_ is a String, then
            1. Let _xStr_ be _x_.
          1. Else if _x_ is an Object, then
            1. Let _code_ be HostGetCodeForEval(_x_).
            1. If _code_ is a String, let _xStr_ be _code_.
            1. Else, return _x_.
          1. Else,
            1. Return _x_.
```
Also possible:
```suggestion
          1. If _x_ is an Object, let _xStr_ be HostGetCodeForEval(_x_). Otherwise, let _xStr_ be _x_.
          1. If _xStr_ is not a String, return _x_.
```

gibson042 replied:
Note also that a later use of _x_ needs updating:
```diff
-1. Let _script_ be ParseText(StringToCodePoints(_x_), |Script|).
+1. Let _script_ be ParseText(StringToCodePoints(_xStr_), |Script|).
```

gibson042 commented on spec.html:
```diff
@@ -28857,10 +28857,18 @@ <h1>
           1. Let _evalRealm_ be the current Realm Record.
           1. NOTE: In the case of a direct eval, _evalRealm_ is the realm of both the caller of `eval` and of the `eval` function itself.
-          1. Perform ? HostEnsureCanCompileStrings(_evalRealm_, « », _x_, _direct_).
+          1. Let _type_ be ~direct-eval~ if _direct_ is *true*, or ~indirect-eval~ if _direct_ is *false*.
```
```suggestion
          1. If _direct_ is *true*, let _type_ be ~direct-eval~. Otherwise, let _type_ be ~indirect-eval~.
```

nicolo-ribaudo replied:
How do you feel about changing the type of `PerformEval`'s `_direct_` parameter to be `~direct-eval~ | ~indirect-eval~` instead? So that:
- we don't have to convert from boolean to enum
- it's clearer what the caller of `PerformEval` mean then when using "true"/"false"

michaelficarra replied:
This is what I was suggesting @nicolo-ribaudo https://github.com/tc39/ecma262/pull/3294#discussion_r1554364846

lukewarlow replied:
Changed this

michaelficarra commented on spec.html:
```diff
@@ -28857,10 +28857,18 @@ <h1>
-          1. If _x_ is not a String, return _x_.
+          1. If _x_ is an Object, then
+            1. Let _code_ be HostGetCodeForEval(_x_).
+            1. If _code_ is ~unknown~ return _x_.
```
Maybe if the "code-like" terminology is going to persist, do

```suggestion
            1. If _code_ is ~not-code-like~ return _x_.
```

nicolo-ribaudo replied:
The name "code-like" is currently not referenced anywhere (it was previously the name of the slot), not even in the TT spec. Maybe just `~no-code~`?

michaelficarra replied:
Yeah I'm happy with that. If there's terminology used in the trusted types spec that you'd like to appropriate, that would also work for me.

jmdyck commented on spec.html:
```diff
@@ -28954,11 +28963,28 @@ <h1>
+        </h1>
+        <dl class="header">
+          <dt>description</dt>
+          <dd>It allows host environments to return a String of code from _argument_ to be used by eval,
+          rather than eval returning _argument_.</dd>
```
Join up into one line.

jmdyck commented on spec.html:
```diff
@@ -28862,30 +28862,37 @@ <h1>eval ( _x_ )</h1>
         <emu-alg>
-          1. Assert: If _direct_ is *false*, then _strictCaller_ is also *false*.
-          1. If _x_ is not a String, return _x_.
+          1. Assert: If _direct_ is *indirect-eval*, then _strictCaller_ is *false*.
```
```suggestion
          1. Assert: If _direct_ is ~indirect-eval~, then _strictCaller_ is *false*.
```

lukewarlow replied:
Good catch

nicolo-ribaudo commented:
@bakkot I was re-reading the `new Function` spec -- you mentioned that the _sourceText_ created in CreateDynamicFunction is not exposed. However, my reading is that step 4 of OrdinaryFunctionCreate stores it in `F.[[SourceText]]`, and then `Function.prototype.toString` will return it.

michaelficarra commented:
@nicolo-ribaudo You may be referring to what I said, and I think I may have confused dynamic functions with built-in functions. While the exact whitespace/comments/etc in the String returned by `Function.prototype.toString` for built-ins is intentionally underspecified, the String returned for dynamic functions is indeed fully specified. If you want, it seems okay to pass that to the host AO. Sorry for the mistake.

syg commented:
> @syg I was reading https://github.com/tc39/proposal-source-phase-imports/issues/59 — should we replace the internal slot on the host-provided object with a HostEvalShouldStringifyObject hook for the same reason?

Is this a comment referencing an earlier version of the PR? I don't see an internal slot currently.

lukewarlow commented:
@syg yes there was previously a slot but we've changed it to a host hook (which itself may or may not check a slot)

lukewarlow commented:
> If you want, it seems okay to pass that to the host AO. Sorry for the mistake.

@michaelficarra is that something we can do without needing to represent that change? Passing the built code string through would indeed be the best course from our side (CSP is in fact already specced that way).

michaelficarra commented:
@lukewarlow Any normative change at stage 2.7 or later should be presented to the committee, but I don't think this one should require consensus. It can be just a quick update.

lukewarlow committed:
Normative: Allow codelike object evaluation.

This change introduces a new HostGetCodeForEval host hook, this is used to allow hosts to evaluate certain objects.

This change also provides the full code string, the compilation type and the original arguments, to the host hook HostEnsureCanCompileStrings.

lukewarlow committed:
format

nicolo-ribaudo commented:
Removing the [needs test262 tests] label per the April 2024 discussion (https://github.com/tc39/notes/blob/main/meetings/2024-04/april-10.md#evalnew-function-changes-for-trusted-types-as-normative-pr-or-stage-3) (the change is not easily testable in test262 since it's just about how much info is exposed to hosts, and it's covered by WPT)

---

Regarding the esmeta failure, it seems like it's checking updated call sites to `HostEnsureCanCompileStrings` using the old signature of the host hook, rather than the new one from this PR?

nicolo-ribaudo commented on spec.html:
```diff
@@ -29891,20 +29898,39 @@ <h1>
+        </h1>
+        <dl class="header">
+          <dt>description</dt>
+          <dd>It allows host environments to return a String of code from _argument_ to be used by eval, rather than eval returning _argument_.</dd>
```
```suggestion
          <dd>It allows host environments to return a String representing _argument_ to be used as code by eval, rather than eval returning _argument_ as-is.</dd>
```

gibson042 replied:
I think this should link to [eval](https://tc39.es/ecma262/multipage/global-object.html#sec-eval-x).
```suggestion
          <dd>It allows host environments to return a String representing _argument_ for <emu-xref href="#sec-eval-x" title></emu-xref> to interpret as source text rather than returning _argument_ as-is.</dd>
```

nicolo-ribaudo commented on spec.html:
```diff
@@ -29891,20 +29898,39 @@ <h1>
+          <dt>description</dt>
+          <dd>It allows host environments to return a String of code from _argument_ to be used by eval, rather than eval returning _argument_.</dd>
+        </dl>
+        <p>_argument_ represents the Object to be checked for code.</p>
```
I would remove this, it's repeating what the `<dl>` description says.

nicolo-ribaudo commented on spec.html:
```diff
@@ -29891,20 +29898,39 @@ <h1>
         <p>
-          _parameterStrings_ represents the strings that, when using one of the function constructors, will be concatenated together to build the parameters list. _bodyString_ represents the function body or the string passed to an `eval` call.
-          _direct_ signifies whether the evaluation is a direct eval.
+          _parameterStrings_ represents the strings that, when using one of the function constructors, will be concatenated together to build the parameters list.
+          _bodyString_ represents the function body or the string passed to an `eval` call.
+          _codeString_ represents the compiled string for a Function or the string passed to an `eval` call.
+          _parameterArgs_ are the values passed as the leading parameters to one of the Function constructors. _bodyArg_ is either the final parameter passed to one of the Function constructors or the value passed to an `eval` call.
```
This list is missing `_compilationType_`.

Also, it's has so many entries now that maybe it should be a `<ol>` list, even though that's slightly unusual. @tc39/ecma262-editors thoughts?

gibson042 replied:
> This list is missing `_compilationType_`.

It is also missing _calleeRealm_.

nicolo-ribaudo reviewed (changes requested):
This PR introduces `_sourceStr_` and `_codeString_` variables: we'll want to use wither `String` or `Str` consistently, but we need to wait for https://github.com/tc39/ecma262/pull/3837 to decide which one.

--

For other reviewers:
- the implementation of `HostEnsureCanCompileStrings` is at https://w3c.github.io/webappsec-csp/#can-compile-strings
- the implementation of `HostGetCodeForEval` is at https://html.spec.whatwg.org/#hostgetcodeforeval(argument)

nicolo-ribaudo commented:
@jhnaldo This PR will need a corresponding update in https://github.com/es-meta/esmeta/blob/main/src/main/resources/manuals/funcs/HostEnsureCanCompileStrings.ir.

Should we add this to `esmeta-ignore.json` for now, so that ESMeta can be updated after that this PR is merged, or is it possible somehow for ESMeta to typecheck both this PR and what's currently on `main`?

gibson042 commented on spec.html:
```diff
@@ -29808,30 +29808,37 @@ <h1>eval ( _source_ )</h1>
-          1. Assert: If _direct_ is *false*, then _strictCaller_ is also *false*.
-          1. If _source_ is not a String, return _source_.
+          1. Assert: If _direct_ is ~indirect-eval~, then _strictCaller_ is *false*.
+          1. If _source_ is a String, then
+            1. Let _sourceStr_ be _source_.
+          1. Else if _source_ is an Object, then
+            1. Let _code_ be HostGetCodeForEval(_source_).
+            1. If _code_ is a String, let _sourceStr_ be _code_.
+            1. Else, return _source_.
+          1. Else,
+            1. Return _source_.
```
Suggested simplification:
```suggestion
          1. If _source_ is an Object, let _sourceStr_ be HostGetCodeForEval(_source_); else let _sourceStr_ be _source_.
          1. If _sourceStr_ is not a String, return _source_.
```

gibson042 commented on spec.html:
```diff
@@ -29891,20 +29898,39 @@ <h1>
-            _direct_: a Boolean,
+            _codeString_: a String,
+            _compilationType_: ~direct-eval~, ~indirect-eval~, or ~function~,
+            _parameterArgs: a List of ECMAScript language values,
```
Typo fix:
```suggestion
            _parameterArgs_: a List of ECMAScript language values,
```

gibson042 commented on spec.html:
```diff
@@ -29891,20 +29898,39 @@ <h1>
         </h1>
         <dl class="header">
           <dt>description</dt>
           <dd>It allows host environments to block certain ECMAScript functions which allow developers to interpret and evaluate strings as ECMAScript code.</dd>
```
Suggested rewording (mostly because "ECMAScript functions" seems more like it refers to [function objects](https://tc39.es/ecma262/multipage/ecmascript-data-types-and-values.html#function-object) than to behaviors of implementations):
```suggestion
          <dd>It allows host environments to block certain ECMAScript facilities which allow String values to be interpreted and evaluated as ECMAScript code.</dd>
```
